src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+
+
+pub struct SqliteContractDatabase {
+    conn: Option<Connection>,
+    save_point: u16
+}
+
+pub struct SqliteDataMap {
+    map_identifier: i64,
+    key_type: TypeSignature,
+    value_type: TypeSignature
+}
+
+impl SqliteContractDatabase {
+    pub fn initialize(filename: &str) -> Result<SqliteContractDatabase> {
+        let mut contract_db = SqliteContractDatabase::inner_open(filename)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS maps_table
+                      (map_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       map_name TEXT,
+                       key_type TEXT,
+                       value_type TEXT)",
+                            NO_PARAMS)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS data_table
+                      (data_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       map_identifier INTEGER,
+                       key TEXT,
+                       value TEXT)",
+                            NO_PARAMS)?;
+        contract_db.execute("CREATE TABLE IF NOT EXISTS contracts
+                      (contract_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT,
+                       contract_data TEXT)",
+                            NO_PARAMS)?;
+
+        contract_db.check_schema()?;
+
+        Ok(contract_db)
+    }
+
+    pub fn open(filename: &str) -> Result<SqliteContractDatabase> {
+        let contract_db = SqliteContractDatabase::inner_open(filename)?;
+
+        contract_db.check_schema()?;
+        Ok(contract_db)
+    }
+
+    fn check_schema(&self) -> Result<()> {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        if let Some(ref conn) = self.conn {
+            let _: String = conn.query_row(sql, &["maps_table"],
+                                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["contracts"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            let _: String = conn.query_row(sql, &["data_table"],
+                           |row| row.get(0))
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+            Ok(())
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+
+    fn inner_open(filename: &str) -> Result<SqliteContractDatabase> {
+        let conn = Connection::open(filename)
+            .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))?;
+        Ok(SqliteContractDatabase {
+            conn: Some(conn),
+            save_point: 0
+        })
+    }
+
+    fn load_map(&self, contract_name: &str, map_name: &str) -> Result<SqliteDataMap> {
```

Comment:
What is `contract_name` here?  If it's a transaction ID, we have a `Txid` struct you could use.

Reply:
It's been left as a string for now, since I don't believe we've spec'ed out how contracts should be referenced yet.